File: apps/web/src/lib/mobile-app-config.ts
import { apiFetch, resolveApiBaseUrlForHostname } from "@vedamatch/api-client";
import { readFile } from "node:fs/promises";
import path from "node:path";

export type MobileAppConfig = {
  iosUrl: string;
  androidUrl: string;
  iosVersion: string;
  androidVersion: string;
};

const MOBILE_APP_CONFIG_TIMEOUT_MS = 1500;
const EMPTY_CONFIG: MobileAppConfig = {
  iosUrl: "",
  androidUrl: "",
  iosVersion: "",
  androidVersion: "",
};

export async function getMobileAppConfig(hostname: string): Promise<MobileAppConfig> {
  const localVersions = await readLocalMobileVersions();
  const envFallback = readEnvMobileAppConfig();
  const baseUrl = resolveApiBaseUrlForHostname(hostname);
  if (!baseUrl) {
    return mergeMobileAppConfig(envFallback, localVersions);
  }

  try {
    const response = await apiFetch<Partial<MobileAppConfig>>(baseUrl, "/mobile-app/config", {
      cache: "no-store",
      signal: AbortSignal.timeout(MOBILE_APP_CONFIG_TIMEOUT_MS),
    });

    return mergeMobileAppConfig(
      {
        iosUrl: String(response.iosUrl || "").trim(),
        androidUrl: String(response.androidUrl || "").trim(),
        iosVersion: String(response.iosVersion || "").trim(),
        androidVersion: String(response.androidVersion || "").trim(),
      },
      localVersions,
      envFallback,
    );
  } catch {
    return mergeMobileAppConfig(envFallback, localVersions);
  }
}

function mergeMobileAppConfig(...configs: Partial<MobileAppConfig>[]): MobileAppConfig {
  return configs.reduce<MobileAppConfig>(
    (acc, config) => ({
      iosUrl: acc.iosUrl || String(config.iosUrl || "").trim(),
      androidUrl: acc.androidUrl || String(config.androidUrl || "").trim(),
      iosVersion: acc.iosVersion || String(config.iosVersion || "").trim(),
      androidVersion: acc.androidVersion || String(config.androidVersion || "").trim(),
    }),
    { ...EMPTY_CONFIG },
  );
}

function readEnvMobileAppConfig(): MobileAppConfig {
  return {
    iosUrl: readEnvValue("SUPPORT_DOWNLOAD_IOS_URL", "NEXT_PUBLIC_SUPPORT_DOWNLOAD_IOS_URL"),
    androidUrl: readEnvValue("SUPPORT_DOWNLOAD_ANDROID_URL", "NEXT_PUBLIC_SUPPORT_DOWNLOAD_ANDROID_URL"),
    iosVersion: "",
    androidVersion: "",
  };
}

function readEnvValue(...keys: string[]): string {
  for (const key of keys) {
    const value = String(process.env[key] || "").trim();
    if (value) {
      return value;
    }
  }
  return "";
}

async function readLocalMobileVersions(): Promise<Partial<MobileAppConfig>> {
  try {
    const [androidContent, iosContent] = await Promise.all([
      readRepoFile("frontend/android/app/build.gradle"),
      readRepoFile("frontend/ios/vedamatch.xcodeproj/project.pbxproj"),
    ]);

    return {
      androidVersion: parseAndroidAppVersion(androidContent),
      iosVersion: parseIOSAppVersion(iosContent),
    };
  } catch {
    return EMPTY_CONFIG;
  }
}

async function readRepoFile(relativePath: string): Promise<string> {
  const repoRoot = path.resolve(process.cwd(), "..");
  return readFile(path.join(repoRoot, relativePath), "utf8");
}

function parseAndroidAppVersion(content: string): string {
  const versionName = extractLastMatch(content, /versionName\s+["']?([^"'\r\n]+)["']?/g);
  const versionCode = extractLastMatch(content, /versionCode\s+(\d+)/g);
  return formatVersionLabel(versionName, versionCode);
}

function parseIOSAppVersion(content: string): string {
  const marketingVersion = extractLastMatch(content, /MARKETING_VERSION = ([^;]+);/g);
  const projectVersion = extractLastMatch(content, /CURRENT_PROJECT_VERSION = ([^;]+);/g);
  return formatVersionLabel(marketingVersion, projectVersion);
}

function extractLastMatch(content: string, pattern: RegExp): string {
  const matches = Array.from(content.matchAll(pattern));
  for (let idx = matches.length - 1; idx >= 0; idx -= 1) {
    const value = String(matches[idx][1] || "").trim().replace(/^['"]|['"]$/g, "");
    if (value) {
      return value;
    }
  }
  return "";
}

function formatVersionLabel(version: string, build: string): string {
  const normalizedVersion = version.trim();
  const normalizedBuild = build.trim();

  if (normalizedVersion && normalizedBuild) {
    return `${normalizedVersion} (${normalizedBuild})`;
  }
  if (normalizedVersion) {
    return normalizedVersion;
  }
  if (normalizedBuild) {
    return normalizedBuild;
  }
  return "";
}

File: server/internal/services/mobile_app_config.go
package services

import (
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"

	"rag-agent-server/internal/database"
	"rag-agent-server/internal/models"
)

type PublicMobileAppConfig struct {
	IOSURL         string                     `json:"iosUrl"`
	AndroidURL     string                     `json:"androidUrl"`
	IOSVersion     string                     `json:"iosVersion"`
	AndroidVersion string                     `json:"androidVersion"`
	AndroidRelease PublicAndroidReleaseConfig `json:"androidRelease"`
}

type PublicAndroidReleaseConfig struct {
	DownloadURL                 string `json:"downloadUrl"`
	AppVersion                  string `json:"appVersion"`
	VersionCode                 int    `json:"versionCode"`
	ReleaseNotes                string `json:"releaseNotes"`
	InstallInstructions         string `json:"installInstructions"`
	MinimumSupportedVersionCode int    `json:"minimumSupportedVersionCode"`
	PublishedAt                 string `json:"publishedAt"`
}

var (
	androidVersionNamePattern  = regexp.MustCompile(`versionName\s+["']?([^"'\r\n]+)["']?`)
	androidVersionCodePattern  = regexp.MustCompile(`versionCode\s+(\d+)`)
	iosMarketingVersionPattern = regexp.MustCompile(`MARKETING_VERSION = ([^;]+);`)
	iosProjectVersionPattern   = regexp.MustCompile(`CURRENT_PROJECT_VERSION = ([^;]+);`)
)

func GetPublicMobileAppConfig() PublicMobileAppConfig {
	androidURL := strings.TrimSpace(getSystemSettingOrEnvOrDB("SUPPORT_DOWNLOAD_ANDROID_URL"))
	appVersion := strings.TrimSpace(getSystemSettingOrEnvOrDB("ANDROID_TESTERS_APP_VERSION"))
	if appVersion == "" {
		appVersion = readAndroidAppVersion()
	}

	return PublicMobileAppConfig{
		IOSURL:         strings.TrimSpace(getSystemSettingOrEnvOrDB("SUPPORT_DOWNLOAD_IOS_URL")),
		AndroidURL:     androidURL,
		IOSVersion:     readIOSAppVersion(),
		AndroidVersion: readAndroidAppVersion(),
		AndroidRelease: PublicAndroidReleaseConfig{
			DownloadURL:                 androidURL,
			AppVersion:                  appVersion,
			VersionCode:                 parsePositiveInt(getSystemSettingOrEnvOrDB("ANDROID_TESTERS_VERSION_CODE")),
			ReleaseNotes:                strings.TrimSpace(getSystemSettingOrEnvOrDB("ANDROID_TESTERS_RELEASE_NOTES")),
			InstallInstructions:         strings.TrimSpace(getSystemSettingOrEnvOrDB("ANDROID_TESTERS_INSTALL_INSTRUCTIONS")),
			MinimumSupportedVersionCode: parsePositiveInt(getSystemSettingOrEnvOrDB("ANDROID_TESTERS_MIN_SUPPORTED_VERSION_CODE")),
			PublishedAt:                 strings.TrimSpace(getSystemSettingOrEnvOrDB("ANDROID_TESTERS_PUBLISHED_AT")),
		},
	}
}

func readAndroidAppVersion() string {
	content, err := readRepoFile("frontend/android/app/build.gradle")
	if err != nil {
		return ""
	}
	return parseAndroidAppVersion(content)
}

func readIOSAppVersion() string {
	content, err := readRepoFile("frontend/ios/vedamatch.xcodeproj/project.pbxproj")
	if err != nil {
		return ""
	}
	return parseIOSAppVersion(content)
}

func parseAndroidAppVersion(content string) string {
	versionName := extractLastSubmatch(androidVersionNamePattern, content)
	versionCode := extractLastSubmatch(androidVersionCodePattern, content)
	return formatVersionLabel(versionName, versionCode)
}

func parseIOSAppVersion(content string) string {
	marketingVersion := extractLastSubmatch(iosMarketingVersionPattern, content)
	projectVersion := extractLastSubmatch(iosProjectVersionPattern, content)
	return formatVersionLabel(marketingVersion, projectVersion)
}

func extractLastSubmatch(pattern *regexp.Regexp, content string) string {
	matches := pattern.FindAllStringSubmatch(content, -1)
	for idx := len(matches) - 1; idx >= 0; idx-- {
		if len(matches[idx]) < 2 {
			continue
		}
		value := strings.TrimSpace(matches[idx][1])
		value = strings.Trim(value, `"'`)
		if value != "" {
			return value
		}
	}
	return ""
}

func formatVersionLabel(version string, build string) string {
	version = strings.TrimSpace(version)
	build = strings.TrimSpace(build)

	switch {
	case version != "" && build != "":
		return version + " (" + build + ")"
	case version != "":
		return version
	case build != "":
		return build
	default:
		return ""
	}
}

func parsePositiveInt(raw string) int {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return 0
	}
	value := 0
	for _, ch := range raw {
		if ch < '0' || ch > '9' {
			return 0
		}
		value = value*10 + int(ch-'0')
	}
	if value < 0 {
		return 0
	}
	return value
}

func readRepoFile(relativePath string) (string, error) {
	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		return "", os.ErrNotExist
	}

	repoRoot := filepath.Clean(filepath.Join(filepath.Dir(currentFile), "..", "..", ".."))
	content, err := os.ReadFile(filepath.Join(repoRoot, relativePath))
	if err != nil {
		return "", err
	}
	return string(content), nil
}

func getSystemSettingOrEnvOrDB(key string) string {
	var setting models.SystemSetting
	if database.DB != nil {
		if err := database.DB.Where("key = ?", key).First(&setting).Error; err == nil {
			if value := strings.TrimSpace(setting.Value); value != "" {
				return value
			}
		}
	}
	return os.Getenv(key)
}
